File: sdk/go/context.go
package nextloggers

import (
	"context"
	"net/http"
)

// LogContext is the language-neutral request/task context attached to log
// records. Go carries it explicitly through context.Context; this package does
// not emulate goroutine-local storage or derive goroutine identifiers.
type LogContext struct {
	LoggedInUser  map[string]any
	Users         []map[string]any
	Fields        map[string]any
	TraceID       string
	TraceIDs      []string
	SpanID        string
	TraceFlags    byte
	TraceFlagsSet bool
	TraceState    string
	Remote        bool
	HasRemote     bool
	Baggage       map[string]string
	RoutineID     string
	Tags          []string
	Context       []any
	Meta          []any
}

// TraceContext remains as a source-compatible name for tracing-focused callers.
type TraceContext = LogContext

type logContextKey struct{}

func cloneStringMap(source map[string]string) map[string]string {
	if source == nil {
		return nil
	}
	target := make(map[string]string, len(source))
	for key, value := range source {
		target[key] = value
	}
	return target
}

func cloneUserList(source []map[string]any) []map[string]any {
	if source == nil {
		return nil
	}
	target := make([]map[string]any, 0, len(source))
	for _, value := range source {
		target = append(target, cloneContextMap(value))
	}
	return target
}

// cloneLogContext returns a deep, detached copy: every nested collection is
// rebuilt so the result never aliases the input. A non-zero TraceFlags is
// treated as explicitly set.
func cloneLogContext(value LogContext) LogContext {
	return LogContext{
		LoggedInUser:  cloneContextMap(value.LoggedInUser),
		Users:         cloneUserList(value.Users),
		Fields:        cloneContextMap(value.Fields),
		TraceID:       value.TraceID,
		TraceIDs:      append([]string(nil), value.TraceIDs...),
		SpanID:        value.SpanID,
		TraceFlags:    value.TraceFlags,
		TraceFlagsSet: value.TraceFlagsSet || value.TraceFlags != 0,
		TraceState:    value.TraceState,
		Remote:        value.Remote,
		HasRemote:     value.HasRemote,
		Baggage:       cloneStringMap(value.Baggage),
		RoutineID:     value.RoutineID,
		Tags:          append([]string(nil), value.Tags...),
		Context:       cloneContextSlice(value.Context),
		Meta:          cloneContextSlice(value.Meta),
	}
}

// firstNonEmpty returns the first non-empty candidate, or "".
func firstNonEmpty(candidates ...string) string {
	for _, candidate := range candidates {
		if candidate != "" {
			return candidate
		}
	}
	return ""
}

func firstString(values []string) string {
	if len(values) == 0 {
		return ""
	}
	return values[0]
}

// mergedBaggage keeps the nil shape of an absent baggage map: a nil base with
// nothing to add stays nil; otherwise both maps are merged into a new one.
func mergedBaggage(base, overlay map[string]string) map[string]string {
	if base == nil && len(overlay) == 0 {
		return nil
	}
	return mergeMaps(base, overlay)
}

// MergeLogContexts applies canonical scope merge rules: maps merge with the
// inner scope winning; users/context/meta append; trace IDs and tags deduplicate
// while retaining order; and present scalar values replace their parent value.
//
// Both inputs are snapshotted first and a new context is built from their
// parts; neither argument, nor anything it references, is written to.
func MergeLogContexts(base LogContext, patch LogContext) LogContext {
	outer := cloneLogContext(base)
	inner := cloneLogContext(patch)
	// Trace ids accumulate in this order: the outer list, the outer primary id,
	// the inner primary id, then the inner list.
	traceIDs := appendAllUnique(
		appendAllUnique(outer.TraceIDs, outer.TraceID, inner.TraceID),
		inner.TraceIDs...,
	)
	// Trace flags are meaningful even when zero. Explicit TraceFlagsSet is the
	// unambiguous signal; trace/span IDs also identify an OTEL span context whose
	// sampled bit may legitimately be zero.
	innerHasSpanContext := inner.TraceFlagsSet || inner.TraceFlags != 0 ||
		inner.TraceID != "" || inner.SpanID != ""
	return LogContext{
		LoggedInUser:  mergeMaps(outer.LoggedInUser, inner.LoggedInUser),
		Users:         concat(outer.Users, mapSlice(inner.Users, cloneContextMap)),
		Fields:        mergeMaps(outer.Fields, inner.Fields),
		TraceID:       firstNonEmpty(inner.TraceID, outer.TraceID, firstString(traceIDs)),
		TraceIDs:      traceIDs,
		SpanID:        firstNonEmpty(inner.SpanID, outer.SpanID),
		TraceFlags:    pick(innerHasSpanContext, inner.TraceFlags, outer.TraceFlags),
		TraceFlagsSet: innerHasSpanContext || outer.TraceFlagsSet,
		TraceState:    firstNonEmpty(inner.TraceState, outer.TraceState),
		Remote:        pick(inner.HasRemote, inner.Remote, outer.Remote),
		HasRemote:     inner.HasRemote || outer.HasRemote,
		Baggage:       mergedBaggage(outer.Baggage, inner.Baggage),
		RoutineID:     firstNonEmpty(inner.RoutineID, outer.RoutineID),
		Tags:          appendAllUnique(outer.Tags, inner.Tags...),
		Context:       concat(outer.Context, inner.Context),
		Meta:          concat(outer.Meta, inner.Meta),
	}
}

// MergeLogContext preserves the tracing-focused value-level API introduced by
// the canonical Go SDK while using the complete LogContext merge semantics.
func MergeLogContext(outer, inner TraceContext) TraceContext {
	return MergeLogContexts(outer, inner)
}

func withLogContextSnapshot(parent context.Context, value LogContext) context.Context {
	if parent == nil {
		parent = context.Background()
	}
	return context.WithValue(parent, logContextKey{}, cloneLogContext(value))
}

// WithLogContext returns a child context carrying an immutable merged snapshot.
// Child goroutines must receive the returned context explicitly.
func WithLogContext(parent context.Context, value LogContext) context.Context {
	if parent == nil {
		parent = context.Background()
	}
	current, _ := LogContextFrom(parent)
	return withLogContextSnapshot(parent, MergeLogContexts(current, value))
}

// WithMergedLogContext is an explicit alias for callers that want the merge
// semantics to be visible at the call site.
func WithMergedLogContext(parent context.Context, patch LogContext) context.Context {
	return WithLogContext(parent, patch)
}

// LogContextFrom returns an immutable copy of the context value.
func LogContextFrom(ctx context.Context) (LogContext, bool) {
	if ctx == nil {
		return LogContext{}, false
	}
	value, ok := ctx.Value(logContextKey{}).(LogContext)
	if !ok {
		return LogContext{}, false
	}
	return cloneLogContext(value), true
}

// UpdateLogContext replaces the current immutable snapshot with the value
// returned by update. The callback receives a defensive copy.
func UpdateLogContext(ctx context.Context, update func(LogContext) LogContext) context.Context {
	if update == nil {
		return ctx
	}
	current, _ := LogContextFrom(ctx)
	return withLogContextSnapshot(ctx, update(current))
}

func WithLoggedInUser(ctx context.Context, user map[string]any) context.Context {
	return WithLogContext(ctx, LogContext{LoggedInUser: user})
}

// LoggedInUserFrom returns a defensive copy of the current authenticated user.
func LoggedInUserFrom(ctx context.Context) (map[string]any, bool) {
	value, ok := LogContextFrom(ctx)
	if !ok || len(value.LoggedInUser) == 0 {
		return nil, false
	}
	return cloneContextMap(value.LoggedInUser), true
}

func WithTraceFlags(ctx context.Context, traceFlags byte) context.Context {
	return WithLogContext(ctx, LogContext{
		TraceFlags:    traceFlags,
		TraceFlagsSet: true,
	})
}

func WithTrace(ctx context.Context, traceID, spanID string) context.Context {
	return WithLogContext(ctx, LogContext{
		TraceID:  traceID,
		TraceIDs: []string{traceID},
		SpanID:   spanID,
	})
}

// RequestWithLogContext snapshots value into a cloned *http.Request.
func RequestWithLogContext(request *http.Request, value LogContext) *http.Request {
	if request == nil {
		return nil
	}
	return request.WithContext(WithLogContext(request.Context(), value))
}

func LogContextFromRequest(request *http.Request) (LogContext, bool) {
	if request == nil {
		return LogContext{}, false
	}
	return LogContextFrom(request.Context())
}

// LogContextMiddleware resolves a request context once and installs it on the
// request passed to downstream handlers. Request.Context remains the canonical
// propagation mechanism across goroutines and cancellation boundaries.
func LogContextMiddleware(resolve func(*http.Request) LogContext) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
			if resolve == nil {
				next.ServeHTTP(writer, request)
				return
			}
			next.ServeHTTP(writer, RequestWithLogContext(request, resolve(request)))
		})
	}
}

// ApplyContext enriches an event from context.Context without retaining the
// context itself or any mutable map supplied by the caller.
func (event *Event) ApplyContext(ctx context.Context) *Event {
	value, ok := LogContextFrom(ctx)
	if !ok {
		return event
	}

	// value is already a detached snapshot, so its fields can be merged with
	// the span attributes into one new map without a second clone.
	fields := mergeMaps(value.Fields, mapOf(
		kvWhen[string, any](value.SpanID != "", "otel.span_id", value.SpanID),
		kvWhen[string, any](value.TraceFlagsSet, "otel.trace_flags", value.TraceFlags),
		kvWhen[string, any](value.TraceState != "", "otel.trace_state", value.TraceState),
		kvWhen[string, any](value.HasRemote, "otel.remote", value.Remote),
		kvWhen[string, any](len(value.Baggage) > 0, "otel.baggage", cloneStringMap(value.Baggage)),
	))
	if len(fields) > 0 {
		event.AddFields(fields)
	}
	if len(value.LoggedInUser) > 0 {
		event.AddLoggedInUserInfo(value.LoggedInUser)
	}
	for _, user := range value.Users {
		event.AddUserInfo(user)
	}
	if value.TraceID != "" {
		// An explicitly attached event trace remains primary; the ambient trace
		// is still retained in TraceIDs for correlation.
		event.AddTrace(value.TraceID)
	}
	for _, traceID := range value.TraceIDs {
		event.AddTrace(traceID)
	}
	if value.RoutineID != "" {
		event.AddRoutineID(value.RoutineID)
	}
	otelContext := value.TraceID != "" || value.SpanID != "" || value.TraceFlagsSet || value.TraceState != "" || value.HasRemote || len(value.Baggage) > 0
	if otelContext {
		event.AddTags("otel")
	}
	if len(value.Tags) > 0 {
		event.AddTags(value.Tags...)
	}
	for _, item := range value.Context {
		event.AddContext(item)
	}
	for _, item := range value.Meta {
		event.AddMeta(item)
	}
	return event
}

File: sdk/rust/src/context.rs
//! Explicit thread/task context for the Rust SDK.
//!
//! Synchronous scopes use a guarded thread-local stack. Async code uses
//! [`ContextFuture`], which installs one immutable snapshot for each poll so a
//! task remains correlated when an executor moves it between threads. Nothing
//! patches an executor, future, runtime, or OpenTelemetry global.

use crate::{Event, JsonObject, Logger, LoggerError, Value};
use std::cell::RefCell;
use std::collections::BTreeMap;
use std::future::Future;
use std::marker::PhantomData;
use std::pin::Pin;
use std::rc::Rc;
use std::task::{Context as TaskContext, Poll};
use std::thread::{self, ThreadId};

#[derive(Clone, Debug, Default, PartialEq)]
pub struct LogContext {
    pub logged_in_user: JsonObject,
    pub users: Vec<JsonObject>,
    pub fields: JsonObject,
    pub trace_id: Option<String>,
    pub trace_ids: Vec<String>,
    pub span_id: Option<String>,
    pub trace_flags: u8,
    pub trace_state: Option<String>,
    pub remote: Option<bool>,
    pub baggage: BTreeMap<String, String>,
    pub routine_id: Option<String>,
    pub tags: Vec<String>,
    pub context: Vec<Value>,
    pub meta: Vec<Value>,
}

fn clean(value: Option<String>, maximum: usize) -> Option<String> {
    let value = value.unwrap_or_default().trim().to_string();
    if value.is_empty() {
        None
    } else if value.len() <= maximum {
        Some(value)
    } else {
        Some(value.chars().take(maximum).collect())
    }
}

fn unique_strings(values: impl IntoIterator<Item = String>) -> Vec<String> {
    values.into_iter().fold(Vec::new(), |unique, value| {
        let value = value.trim().to_string();
        if value.is_empty() || unique.contains(&value) {
            unique
        } else {
            unique.into_iter().chain(std::iter::once(value)).collect()
        }
    })
}

impl LogContext {
    pub fn normalized(mut self) -> Self {
        self.trace_id = clean(self.trace_id, 256);
        self.span_id = clean(self.span_id, 256);
        self.trace_state = clean(self.trace_state, 512);
        self.routine_id = clean(self.routine_id, 256);
        self.trace_ids = unique_strings(self.trace_ids.into_iter().chain(self.trace_id.clone()));
        self.tags = unique_strings(self.tags);
        self
    }
}

pub fn merge_log_context(outer: &LogContext, inner: &LogContext) -> LogContext {
    let outer = outer.clone().normalized();
    let inner = inner.clone().normalized();

    let logged_in_user = outer
        .logged_in_user
        .into_iter()
        .chain(inner.logged_in_user.clone())
        .collect();
    let fields = outer
        .fields
        .into_iter()
        .chain(inner.fields.clone())
        .collect();
    let baggage = outer
        .baggage
        .into_iter()
        .chain(inner.baggage.clone())
        .collect();

    let trace_id = inner.trace_id.clone().or(outer.trace_id.clone());
    let trace_ids = unique_strings(
        outer
            .trace_ids
            .into_iter()
            .chain(inner.trace_ids)
            .chain(trace_id.clone()),
    );
    let tags = unique_strings(outer.tags.into_iter().chain(inner.tags));
    let inner_has_span_context = inner.trace_id.is_some()
        || inner.span_id.is_some()
        || inner.trace_state.is_some()
        || inner.remote.is_some()
        || inner.trace_flags != 0;

    LogContext {
        logged_in_user,
        users: outer.users.into_iter().chain(inner.users.clone()).collect(),
        fields,
        trace_id,
        trace_ids,
        span_id: inner.span_id.clone().or(outer.span_id),
        trace_flags: if inner_has_span_context {
            inner.trace_flags
        } else {
            outer.trace_flags
        },
        trace_state: inner.trace_state.clone().or(outer.trace_state),
        remote: inner.remote.or(outer.remote),
        baggage,
        routine_id: inner.routine_id.clone().or(outer.routine_id),
        tags,
        context: outer
            .context
            .into_iter()
            .chain(inner.context.clone())
            .collect(),
        meta: outer.meta.into_iter().chain(inner.meta.clone()).collect(),
    }
    .normalized()
}

thread_local! {
    static CONTEXT_STACK: RefCell<Vec<LogContext>> = const { RefCell::new(Vec::new()) };
}

pub fn current_log_context() -> LogContext {
    CONTEXT_STACK.with(|stack| stack.borrow().last().cloned().unwrap_or_default())
}

fn push_exact(value: LogContext) -> LogContextGuard {
    let thread_id = thread::current().id();
    let depth = CONTEXT_STACK.with(|stack| {
        let mut stack = stack.borrow_mut();
        let depth = stack.len();
        stack.push(value.normalized());
        depth
    });
    LogContextGuard {
        depth,
        thread_id,
        active: true,
        _not_send: PhantomData,
    }
}

pub fn enter_log_context(value: LogContext) -> LogContextGuard {
    let merged = merge_log_context(&current_log_context(), &value);
    push_exact(merged)
}

#[must_use = "dropping the guard exits the logging context"]
pub struct LogContextGuard {
    depth: usize,
    thread_id: ThreadId,
    active: bool,
    // A scope cannot be moved to or closed from another thread.
    _not_send: PhantomData<Rc<()>>,
}

impl LogContextGuard {
    pub fn close(mut self) {
        self.restore();
    }

    fn restore(&mut self) {
        if !self.active {
            return;
        }
        debug_assert_eq!(self.thread_id, thread::current().id());
        CONTEXT_STACK.with(|stack| stack.borrow_mut().truncate(self.depth));
        self.active = false;
    }
}

impl Drop for LogContextGuard {
    fn drop(&mut self) {
        self.restore();
    }
}

pub fn with_log_context<T>(value: LogContext, callback: impl FnOnce() -> T) -> T {
    let _scope = enter_log_context(value);
    callback()
}

pub fn capture_log_context() -> LogContext {
    current_log_context()
}

pub fn with_captured_log_context<T>(captured: &LogContext, callback: impl FnOnce() -> T) -> T {
    let _scope = push_exact(captured.clone());
    callback()
}

pub fn update_log_context(
    update: impl FnOnce(LogContext) -> LogContext,
) -> Result<LogContext, LoggerError> {
    CONTEXT_STACK.with(|stack| {
        let mut stack = stack.borrow_mut();
        let current = stack
            .last()
            .cloned()
            .ok_or_else(|| LoggerError("next_loggers: no active log context to update".into()))?;
        let updated = update(current).normalized();
        *stack
            .last_mut()
            .expect("active context disappeared while updating") = updated.clone();
        Ok(updated)
    })
}

pub struct ContextFuture<F> {
    context: LogContext,
    future: Pin<Box<F>>,
}

impl<F> ContextFuture<F> {
    pub fn new(context: LogContext, future: F) -> Self {
        Self {
            context: merge_log_context(&current_log_context(), &context),
            future: Box::pin(future),
        }
    }
}

impl<F: Future> Future for ContextFuture<F> {
    type Output = F::Output;

    fn poll(mut self: Pin<&mut Self>, task: &mut TaskContext<'_>) -> Poll<Self::Output> {
        let context = self.context.clone();
        with_captured_log_context(&context, || self.future.as_mut().poll(task))
    }
}

pub fn with_log_context_async<F: Future>(context: LogContext, future: F) -> ContextFuture<F> {
    ContextFuture::new(context, future)
}

pub fn contextualize_future<F: Future>(context: LogContext, future: F) -> ContextFuture<F> {
    with_log_context_async(context, future)
}

pub fn apply_log_context(event: Event, context: &LogContext) -> Event {
    let context = context.clone().normalized();
    let event = match context.trace_id {
        Some(trace_id) => event.add_trace(trace_id, true),
        None => event,
    };
    let event = context
        .trace_ids
        .into_iter()
        .fold(event, |event, trace_id| event.add_trace(trace_id, false));

    let span_fields = [
        context
            .span_id
            .map(|span_id| ("otel.span_id".into(), Value::String(span_id))),
        Some((
            "otel.trace_flags".into(),
            Value::Number(context.trace_flags.into()),
        )),
        context
            .trace_state
            .map(|trace_state| ("otel.trace_state".into(), Value::String(trace_state))),
        context
            .remote
            .map(|remote| ("otel.remote".into(), Value::Bool(remote))),
        (!context.baggage.is_empty()).then(|| {
            (
                "otel.baggage".into(),
                Value::Object(
                    context
                        .baggage
                        .into_iter()
                        .map(|(key, value)| (key, Value::String(value)))
                        .collect(),
                ),
            )
        }),
    ];
    let fields =
        span_fields
            .into_iter()
            .flatten()
            .fold(context.fields, |mut fields, (key, value)| {
                fields.insert(key, value);
                fields
            });
    let event = event.add_fields(fields);
    let event = if context.logged_in_user.is_empty() {
        event
    } else {
        event.add_logged_in_user_info(context.logged_in_user)
    };
    let event = context.users.into_iter().fold(event, Event::add_user_info);
    let event = match context.routine_id {
        Some(routine_id) => event.add_routine_id(routine_id),
        None => event,
    };
    let event = event.add_tags(std::iter::once("otel".to_string()).chain(context.tags));
    let event = context.context.into_iter().fold(event, Event::add_context);
    context.meta.into_iter().fold(event, Event::add_meta)
}

pub fn apply_current_log_context(event: Event) -> Event {
    apply_log_context(event, &current_log_context())
}

impl Logger {
    pub fn trace_context(&self, values: Vec<Value>) -> Event {
        apply_current_log_context(self.trace(values))
    }

    pub fn debug_context(&self, values: Vec<Value>) -> Event {
        apply_current_log_context(self.debug(values))
    }

    pub fn info_context(&self, values: Vec<Value>) -> Event {
        apply_current_log_context(self.info(values))
    }

    pub fn warn_context(&self, values: Vec<Value>) -> Event {
        apply_current_log_context(self.warn(values))
    }

    pub fn error_context(&self, values: Vec<Value>) -> Event {
        apply_current_log_context(self.error(values))
    }

    pub fn fatal_context(&self, values: Vec<Value>) -> Event {
        apply_current_log_context(self.fatal(values))
    }
}
